src/cargo/core/compiler/custom_build.rs

```diff
@@ -249,6 +249,12 @@ fn build_work(cx: &mut Context<'_, '_>, unit: &Unit) -> CargoResult<Job> {
         }
     }
 
+    // Also inform the build script of the rustc compiler context.
+    if let Some(wrapper) = bcx.rustc().wrapper.as_ref() {
+        cmd.env("CARGO_RUSTC_WRAPPER", wrapper);
```

Comment:
Could this (and the one below) perhaps remove the `CARGO_` prefix? They're already well-known env vars for Cargo and we'd just be forcing the env var to be present to match the outer Cargo's configuration, so I'm not sure if we'd gain too much namespacing them.

Reply:
Right now, `RUSTFLAGS` already leaks through to the build script, and I (in `autocfg`) have to apply heuristics to decide whether we're in the right host/target context that those should apply. It would be much simpler for me to consume a *new* name that is unambiguously applicable.

If I must, I could check the versions to know if I'm in this new world, but that will also raise questions about dealing with alternative Rust implementations.

Reply:
Conventionally Cargo mostly sets the normal env vars for things like this, for example it sets `RUSTC` instead of `CARGO_RUSTC`. Since `autocfg` is already dealing with rust version and such it presumably wouldn't be too hard to have a conditional check around the preexisting logic and whether or not to use `RUSTFLAGS`?

Reply:
It would not be difficult. I just feel a little sad about hard-coding version knowledge.

For cases like `RUSTC_WRAPPER` that are conditionally set here, are there cases that would make it necessary to unset it? Like, if an existing `RUSTC_WRAPPER` from outside would not apply in some instance.

Reply:
I'll also add that there _is_ precedent for setting `CARGO_`-prefixed env vars for build scripts, like `CARGO_MANIFEST_LINKS` or `CARGO_CFG_TARGET_ARCH`. They're not _quite_ the same since they aren't also input variables, but I do think it's valuable to not _intentionally_ shadow inputs, and instead leave it to any proxy to turn the out-vars (like `CARGO_RUSTFLAGS`) into in-vars (like `RUSTFLAGS`) as they see fit.

Reply:
Yes Cargo goes both ways. If I could change the design then we'd use `CARGO_` more places than we did today, but my thinking is that if Cargo supports the bare form of the flag then it should continue to use the bare form, but if it's brand new then it should use a `CARGO_` prefix. For example `OUT_DIR` I think should have been prefixed (no changing that now), but these two variables, `RUSTFLAGS` and `RUSTC_WRAPPER` are both supported in their bare form with no prefix (I think rightfully so) so I think we should mirror that to the build script.

Note that this is all if we go this route. If we go a route where `RUSTC` is a script that handles everything then we should scope them so things don't clash.

Reply:
My worry is that non-`CARGO_` _forces_ the Cargo output vars to override variables already in the environment (as @cuviper pointed out) _and_ it means that the vars will _implicitly_ affect subsequent invocations of `cargo` (or other tools that understand them). With the `CARGO_` prefix, those choices must both be made explicitly, which _seems_ better to me here.

Reply:
That's actually sort of my intention, though, which is to affect tools that otherwise read these variables like `cargo` does. Currently you get inconsistent behavior because if you configured the build via `RUSTFLAGS` then it gets inherited to build scripts, but if you configured it via `.cargo/config.toml` then it doesn't get passed through to build scripts. This means that these tools, if at all, already have to handle this case today.

I think it'd be best if we make it consistently always defined or always-not-defined instead of the inbetween we have today.